File: src/geometry.cpp
#include "dcpwizard/geometry.h"

#include <filesystem>
#include <spdlog/spdlog.h>
#include <sstream>

namespace fs = std::filesystem;

namespace dcpwizard
{

int apply_geometry(const fs::path& input_dir,
                   const fs::path& output_dir,
                   const GeometryConfig& config)
{
  if (!fs::exists(input_dir))
  {
    spdlog::error("Input directory not found: {}", input_dir.string());
    return 1;
  }
  if (config.mode == ScaleMode::None)
  {
    spdlog::info("No geometry transform needed");
    return 0;
  }

  fs::create_directories(output_dir);

  // Build ffmpeg video filter
  std::string vf;
  switch (config.mode)
  {
    case ScaleMode::Scale:
      vf = "scale=" + std::to_string(config.target_width) + ":" +
           std::to_string(config.target_height);
      break;
    case ScaleMode::Crop:
    {
      std::ostringstream ss;
      ss << "crop=iw-" << (config.crop_left + config.crop_right)
         << ":ih-" << (config.crop_top + config.crop_bottom)
         << ":" << config.crop_left << ":" << config.crop_top;
      vf = ss.str();
      break;
    }
    case ScaleMode::Letterbox:
      vf = "scale=" + std::to_string(config.target_width) + ":" +
           std::to_string(config.target_height) +
           ":force_original_aspect_ratio=decrease,pad=" +
           std::to_string(config.target_width) + ":" +
           std::to_string(config.target_height) + ":(ow-iw)/2:(oh-ih)/2";
      break;
    case ScaleMode::PillarBox:
      vf = "scale=" + std::to_string(config.target_width) + ":" +
           std::to_string(config.target_height) +
           ":force_original_aspect_ratio=decrease,pad=" +
           std::to_string(config.target_width) + ":" +
           std::to_string(config.target_height) + ":(ow-iw)/2:(oh-ih)/2";
      break;
    default:
      return 0;
  }

  spdlog::info("Applying geometry: {} ({}x{})", vf, config.target_width, config.target_height);

  std::string cmd = "ffmpeg -y -framerate 24 -i \"" +
                    (input_dir / "frame_%06d.tiff").string() +
                    "\" -vf \"" + vf + "\" -pix_fmt rgb48le \"" +
                    (output_dir / "frame_%06d.tiff").string() +
                    "\" 2>/dev/null";

  int rc = system(cmd.c_str());
  if (rc != 0)
  {
    spdlog::error("ffmpeg geometry transform failed");
    return 1;
  }

  uint32_t count = 0;
  for (const auto& entry : fs::directory_iterator(output_dir))
    if (entry.path().extension() == ".tiff")
      ++count;

  spdlog::info("Geometry applied: {} frames in {}", count, output_dir.string());
  return 0;
}

} // namespace dcpwizard

File: rust/crates/dcpwizard-core/src/geometry.rs
use serde::{Deserialize, Serialize};
use std::path::PathBuf;

/// Scaling/framing mode.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, Serialize, Deserialize)]
pub enum ScaleMode {
    #[default]
    None,
    Scale,
    Crop,
    Letterbox,
    PillarBox,
}

/// Geometry / framing configuration.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct GeometryConfig {
    pub input_file: PathBuf,
    pub output_file: PathBuf,
    pub mode: ScaleMode,
    pub target_width: u32,
    pub target_height: u32,
}

/// Apply geometry transform using ffmpeg scale/crop/pad filters.
pub fn apply_geometry(config: &GeometryConfig) -> i32 {
    if config.target_width == 0 || config.target_height == 0 {
        tracing::error!("Target dimensions must be non-zero");
        return -1;
    }

    let filter = build_filter(config);

    if filter.is_empty() {
        // No transform needed, just copy
        match std::fs::copy(&config.input_file, &config.output_file) {
            Ok(_) => return 0,
            Err(e) => {
                tracing::error!("Failed to copy file: {e}");
                return -1;
            }
        }
    }

    let result = std::process::Command::new("ffmpeg")
        .arg("-y")
        .arg("-i")
        .arg(&config.input_file)
        .arg("-vf")
        .arg(&filter)
        .arg("-c:a")
        .arg("copy")
        .arg(&config.output_file)
        .output();

    match result {
        Ok(o) if o.status.success() => {
            tracing::info!(
                "Applied {:?} geometry to {}",
                config.mode,
                config.output_file.display()
            );
            0
        }
        Ok(o) => {
            tracing::error!(
                "ffmpeg geometry transform failed: {}",
                String::from_utf8_lossy(&o.stderr)
            );
            -1
        }
        Err(e) => {
            tracing::error!("Failed to run ffmpeg: {e}");
            -1
        }
    }
}

fn build_filter(config: &GeometryConfig) -> String {
    let w = config.target_width;
    let h = config.target_height;

    match config.mode {
        ScaleMode::None => String::new(),
        ScaleMode::Scale => {
            format!("scale={w}:{h}:flags=lanczos")
        }
        ScaleMode::Crop => {
            // Scale to cover target, then crop to exact size
            format!("scale={w}:{h}:force_original_aspect_ratio=increase,crop={w}:{h}")
        }
        ScaleMode::Letterbox => {
            // Scale to fit within target, then pad with black bars top/bottom
            format!(
                "scale={w}:{h}:force_original_aspect_ratio=decrease,pad={w}:{h}:(ow-iw)/2:(oh-ih)/2:black"
            )
        }
        ScaleMode::PillarBox => {
            // Scale to fit height, then pad with black bars left/right
            format!(
                "scale={w}:{h}:force_original_aspect_ratio=decrease,pad={w}:{h}:(ow-iw)/2:(oh-ih)/2:black"
            )
        }
    }
}
